### ISSUE TITLE
vl-convert is stuck when saving a spec with images produced by altair_tiles

### ISSUE THREAD

**Issue by binste** [vega/vl-convert#63]
I'm using `0.10.2` and the code from #59 works as expected. However, when I try to render an Altair chart produced by the development version of [altair_tiles](https://github.com/altair-viz/altair_tiles) then the kernel/vl-convert gets stuck:

![image](https://github.com/vega/vl-convert/assets/23366411/7480f314-f0ad-4337-b789-e17dcb7c0186)

As there is no error message I don't know how to debug this. @jonmmease Do you have any idea what could cause this? I can download a PNG from the Vega Editor: [Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAHJQAMlAFYQ2AOxAAaEFBUAzGmTShiNOAHcDmlZhrKGbBhAAScPVjQBmGTI1blVm3YgAdRoAE3p3TwBfSI0QpEwkCDhMCHNEHEwATzQAbWBIgF0YjGcyV1QANk8NQSRMuAAnXNA4hPNlJAQ4NBArQTgAfWEITHUQJIBHBjhlKG7UUESelA0RpAbR1C9xzDYcNABGSWiNGa0Q630FkAYGwXNdOEEQntv7jSyceZBlNgRrJD3YoADwexmePVBqyggPmNkEgg+mS+PSmSF8dHiNFIIGK2Wujwh6Gy0NhaHhiN6yO+aIxCSsOJOIGQDQA1uYoMJ9qhMA1php4C5NqA4MCcE10H1BhAaAAvOAAAgA1AqDrikSj0DRkGRuhoTKFwtdReKelKBjL5crVbjiuK2PI4FArCoOTNMI1cl4xDIChoGmx6fMcvwDl5w37xjD+mgACzVKma5mNGG7JrFXnoiDaNgNBDNEBLdBMMbRqAMWoenqTaazOAACi8qskAEp1aBtJWPapUDkQCWimpFql0G9S4Dy5XvgByLC4fBEKWUPYzEYNODJZA4ZcNMgEafW2VsP4DOY0QTW6f7631+utBgIShIa0XGgDTvHhrWqUQU92XwtgqACkCo-n+DAAZe14qne8QPpQTAvvI8jvoIn6ASq06UDgyhkNO7aFiOIBQpoE4VvE3z3o+z4AFSgee0pyoqKohMCN7mpaioEAqABMbYxMOPQkqRgiThRPRUQhCp0RxTEvpk7EMRacncXxBG6IIHoSiAt6Sc+LE0G+H65t+DG-loEGYIBIFgRZkEAHwALwKjICoAGRuQqsGYPBiEschqHoQqTkue5nm6XB1EvoZgUmSqtn-lZwH0f05mJYBAA8zkJZZYVeZJfkKiEAXGQ0mXZWZ4G+Lig4tPEKAErmyDCiA2hwHB649HZvLGKkGrfLsOAOkoqjFGO6BznghAEFAITKAoEAhE82INJQyjJAQyg4AgvAUPwrSJMkEAAALEAclA8QAnLIBAHQQJi5s8oZho+iiusUZxsBcuFggiYJPC8WovDU5AzIDFL9aiDDolY9LYt0TIsuyBLnoIADyOBIFAdD4nIByrLybCst8uborqYxrkTcDBGEMBoDxkPoLqbAQDASAonaAaOs67A9qAcy+J6vber6-qBuJvahuGaii1GZKoPG2yfN8XQNKmua2kOIAHe0nTfOkWTqiAn3fVcoCgtcxCAtMaAyHi5hW4INvGmK2mCmUozRMUSPmICeg9iA-TaKMsQW24sT4vwkgfKKmw6QAwoBGMzAAyryG6YAAsuzCo+D1TAMGmfWJgNse2pGmMNJ0I55D8es9Ewh0DLJ8pjI7zs8WIFQCcbrs9IrYwdF0PTiha0YI1rJraWhZC3jx0kKgACgAkgvo-QLCgHcY3STN0pnFbwqM-1mpGhD98R4nv0pBApPffoGegj1pfCBDHAN9tmf9foC-p7OLfIp74gCGiYE+agFS-0fp-Ouw9JSVW6gRKeDcm4t0VHREBYCIHHlftfJ4Cp+BYJPFAwe39ej7yYogoB1hSAbHrNOe03MXTKGnOAnITZfTQPPsg3eQ1rCezvqadA9Yd6DF4b4NhBRrRHBkIfNwVQF45Wql-WB2tDKUMEa1NCuY7yGU4aQ18sV0wCO0jooyWivwENfIBGS5DW7KMoqCHuSChEHHwUfNgs8EjKGEU3MRmAcgHEkXRFeCpuLSIwqqUJucWY+J4WwPhASglL1XmEyQMjD4r0PvPGxqUqohxgZRZC6jtKlTvMhPRKjiooVKsUnoAYIIhF0iVcxbiqnWJSoxeUFTKLZEKPqQ0dNKhRCAA)

# Code example

```python
import vl_convert as vlc

# This specification is the same as in the Vega editor link above
tile_chart_spec = r"""{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.8.0.json",
  "config": {
    "view": {
      "continuousHeight": 300,
      "continuousWidth": 300
    }
  },
  "datasets": {
    "empty": [
      {}
    ]
  },
  "height": 600,
  "layer": [
    {
      "data": {
        "name": "tile_list",
        "sequence": {
          "as": "a",
          "start": 0,
          "stop": 18
        }
      },
      "encoding": {
        "url": {
          "field": "url",
          "type": "nominal"
        },
        "x": {
          "field": "x",
          "scale": null,
          "type": "quantitative"
        },
        "y": {
          "field": "y",
          "scale": null,
          "type": "quantitative"
        }
      },
      "mark": {
        "clip": true,
        "height": {
          "expr": "tile_size + 1"
        },
        "type": "image",
        "width": {
          "expr": "tile_size + 1"
        }
      },
      "projection": {
        "center": [
          0,
          50
        ],
        "rotate": [
          -10,
          0,
          0
        ],
        "scale": 400,
        "type": "mercator"
      },
      "transform": [
        {
          "as": "b",
          "calculate": "sequence(0, 18)"
        },
        {
          "flatten": [
            "b"
          ]
        },
        {
          "as": "url",
          "calculate": "'https://tile.openstreetmap.org/' + zoom_ceil + '/' + ((datum.a + dii_floor + tiles_count) % tiles_count) + '/' + (datum.b + djj_floor) + '.png'"
        },
        {
          "as": "x",
          "calculate": "datum.a * tile_size + dx + (tile_size / 2)"
        },
        {
          "as": "y",
          "calculate": "datum.b * tile_size + dy + (tile_size / 2)"
        },
        {
          "filter": "((datum.a + dii_floor + tiles_count) % tiles_count) >= 0 && (datum.b + djj_floor) >= 0 && ((datum.a + dii_floor + tiles_count) % tiles_count) <= tiles_count && (datum.b + djj_floor) <= tiles_count"
        }
      ]
    },
    {
      "data": {
        "format": {
          "feature": "countries",
          "type": "topojson"
        },
        "url": "https://cdn.jsdelivr.net/npm/vega-datasets@v1.29.0/data/world-110m.json"
      },
      "encoding": {
        "fill": {
          "field": "id",
          "legend": null,
          "type": "quantitative"
        }
      },
      "mark": {
        "fillOpacity": 0.1,
        "stroke": "orange",
        "strokeWidth": 2,
        "type": "geoshape"
      },
      "projection": {
        "center": [
          0,
          50
        ],
        "rotate": [
          -10,
          0,
          0
        ],
        "scale": 400,
        "type": "mercator"
      }
    },
    {
      "data": {
        "name": "empty"
      },
      "encoding": {
        "x": {
          "value": 0
        },
        "y": {
          "value": {
            "expr": "height"
          }
        }
      },
      "mark": {
        "align": "left",
        "dx": 3,
        "dy": -8,
        "text": "(C) OpenStreetMap contributors",
        "type": "text"
      }
    }
  ],
  "params": [
    {
      "name": "base_tile_size",
      "value": 256
    },
    {
      "expr": "400",
      "name": "pr_scale"
    },
    {
      "expr": "log((2 * PI * pr_scale) / base_tile_size) / log(2)",
      "name": "zoom_level"
    },
    {
      "expr": "ceil(zoom_level)",
      "name": "zoom_ceil"
    },
    {
      "expr": "pow(2, zoom_ceil)",
      "name": "tiles_count"
    },
    {
      "expr": "base_tile_size * pow(2, zoom_level - zoom_ceil)",
      "name": "tile_size"
    },
    {
      "expr": "invert('projection', [0, 0])",
      "name": "base_point"
    },
    {
      "expr": "(base_point[0] + 180) / 360 * tiles_count",
      "name": "dii"
    },
    {
      "expr": "floor(dii)",
      "name": "dii_floor"
    },
    {
      "expr": "(dii_floor - dii) * tile_size",
      "name": "dx"
    },
    {
      "expr": "(1 - log(tan(base_point[1] * PI / 180) + 1 / cos(base_point[1] * PI / 180)) / PI) / 2 * tiles_count",
      "name": "djj"
    },
    {
      "expr": "floor(djj)",
      "name": "djj_floor"
    },
    {
      "expr": "round((djj_floor - djj) * tile_size)",
      "name": "dy"
    }
  ],
  "width": 600
}
"""
png_data = vlc.vegalite_to_png(tile_chart_spec)
```

**Comment by jonmmease** [comment]
Thanks for the report and repro, I'll take a look

**Comment by jonmmease** [comment]
When I open the example in the Vega editor ([Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAHJQAMlAFYQ2AOxAAaEFBUAzGmTShiNOAHcDmlZhrKGbBhAAScPVjQBmGTI1blVm3YgAdRoAE3p3TwBfSI0QpEwkCDhMCHNEHEwATzQAbWBIgF0YjGcyV1QANk8NQSRMuAAnXNA4hPNlJAQ4NBArQTgAfWEITHUQJIBHBjhlKG7UUESelA0RpAbR1C9xzDYcNABGSWiNGa0Q630FkAYGwXNdOEEQntv7jSyceZBlNgRrJD3YoADwexmePVBqyggPmNkEgg+mS+PSmSF8dHiNFIIGK2Wujwh6Gy0NhaHhiN6yO+aIxCSsOJOIGQDQA1uYoMJ9qhMA1php4C5NqA4MCcE10H1BhAaAAvOAAAgA1AqDrikSj0DRkGRuhoTKFwtdReKelKBjL5crVbjiuK2PI4FArCoOTNMI1cl4xDIChoGmx6fMcvwDl5w37xjD+mgACzVKma5mNGG7JrFXnoiDaNgNBDNEBLdBMMbRqAMWoenqTaazOAACi8qskAEp1aBtJWPapUDkQCWimpFql0G9S4Dy5XvgByLC4fBEKWUPYzEYNODJZA4ZcNMgEafW2VsP4DOY0QTW6f7631+utBgIShIa0XGgDTvHhrWqUQU92XwtgqACkCo-n+DAAZe14qne8QPpQTAvvI8jvoIn6ASq06UDgyhkNO7aFiOIBQpoE4VvE3z3o+z4AFSgee0pyoqKohMCN7mpaioEAqABMbYxMOPQkqRgiThRPRUQhCp0RxTEvpk7EMRacncXxBG6IIHoSiAt6Sc+LE0G+H65t+DG-loEGYIBIFgRZkEAHwALwKjICoAGRuQqsGYPBiEschqHoQqTkue5nm6XB1EvoZgUmSqtn-lZwH0f05mJYBAA8zkJZZYVeZJfkKiEAXGQ0mXZWZ4G+Lig4tPEKAErmyDCiA2hwHB649HZvLGKkGrfLsOAOkoqjFGO6BznghAEFAITKAoEAhE82INJQyjJAQyg4AgvAUPwrSJMkEAAALEAclA8QAnLIBAHQQJi5s8oZho+iiusUZxsBcuFggiYJPC8WovDU5AzIDFL9aiDDolY9LYt0TIsuyBLnoIADyOBIFAdD4nIByrLybCst8uborqYxrkTcDBGEMBoDxkPoLqbAQDASAonaAaOs67A9qAcy+J6vber6-qBuJvahuGaii1GZKoPG2yfN8XQNKmua2kOIAHe0nTfOkWTqiAn3fVcoCgtcxCAtMaAyHi5hW4INvGmK2mCmUozRMUSPmICeg9iA-TaKMsQW24sT4vwkgfKKmw6QAwoBGMzAAyryG6YAAsuzCo+D1TAMGmfWJgNse2pGmMNJ0I55D8es9Ewh0DLJ8pjI7zs8WIFQCcbrs9IrYwdF0PTiha0YI1rJraWhZC3jx0kKgACgAkgvo-QLCgHcY3STN0pnFbwqM-1mpGhD98R4nv0pBApPffoGegj1pfCBDHAN9tmf9foC-p7OLfIp74gCGiYE+agFS-0fp-Ouw9JSVW6gRKeDcm4t0VHREBYCIHHlftfJ4Cp+BYJPFAwe39ej7yYogoB1hSAbHrNOe03MXTKGnOAnITZfTQPPsg3eQ1rCezvqadA9Yd6DF4b4NhBRrRHBkIfNwVQF45Wql-WB2tDKUMEa1NCuY7yGU4aQ18sV0wCO0jooyWivwENfIBGS5DW7KMoqCHuSChEHHwUfNgs8EjKGEU3MRmAcgHEkXRFeCpuLSIwqqUJucWY+J4WwPhASglL1XmEyQMjD4r0PvPGxqUqohxgZRZC6jtKlTvMhPRKjiooVKsUnoAYIIhF0iVcxbiqnWJSoxeUFTKLZEKPqQ0dNKhRCAA)), it takes around 5 seconds for the background image to fill in. Is that expected at the moment?

In the vega editor it looks like `tile_list` dataset has hundreds of rows, do you think all of these images are being loaded by the browser? I'm wondering if vl-convert is loading all of these images serially (rather than in parallel the way a browser might) and so it's just taking a long long long time.

**Comment by binste** [comment]
Thank you for your help! The loading time is indeed not great although I find it to be much faster in a Jupyter notebook. Not sure if/how I can improve that.

Hmm good point. `tile_list` might still contain too many URLs and I try to discard all URLs which are not shown in the chart in the filter transform (Python code is [here](https://github.com/altair-viz/altair_tiles/blob/b2e83cb821c028c3ff338a63509c5b194fcf2778/altair_tiles/__init__.py#L228)). If I display the chart in the browser, Vega seems to only load the images which are shown anyway so I have not investigated if I discard enough or if I could get rid of more. I need to have another look at that logic.

For my understanding, does vl-convert load all images in `tile_list` before the filter is applied or is it converted to an SVG, i.e. after all filters are applied, and then all URLs which are in that SVG are downloaded?

**Comment by binste** [comment]
You're right. This might not be a vl-convert issue but something I should fix on the altair_tiles side. I leave this open until I figured it out if that's ok but no need for you to further investigate.

I tested it with a spec that should just show 2 tiles (I added a space between them so it's visible): [Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAHJQAMlAFYQ2AOxAAaEFBUAzGmTShiNOAHcDmlZhrKGbBhAAScPVjQBmGTI1blVm3YgAdRoAE3p3TwBfSI0QpEwkCDhMCHNEHEwATzQAbWBIgF0YjGcyV1QANk8NQSRMuAAnXNA4hPNlJAQ4NBArQTgAfWEITHUQJIBHBjhlKG7UUESelA0RpAbR1C9xzDYcNAAWaI0ZrRDrfQWQBgbBc104QRCem7uNLJx5kGU2BGskO7FAAe92MTx6INWUAB8xsgkE70ynx6UyQvjo8RopBAxWyVwe4PQ2ShMLQcIRvSRX1R6ISVmxxxAyAaAGtzFBhPtUJgGtMNPAXJtQHAgTgmug+oMIDQAF5wAAE-HlAEYZDjEcj0DRkGRuhoTKFwlcRWKepKBtK5YqVWrGWK2PI4FArCp2TNMI1cl4xDIChoGmw6fMcvxVWovL6Sf1DtVKZqmY1obsmsUeWiINo2A0EM0QEt0EwxtDBFAGLUPT1JtNZnAABReeUHACU6tA2nLHtUqByIELRTUi1S6FeRYBpfLXwA5FhcPgiJLKHsZiMGnBksgcIuGmQCJP5QBqeUyti-AZzGiCA-yye7q+12utBgIShIK-nGgDdsnhpXyUQM92L4TbygApPKf4AQwQFXjee6Hg+8RPpQTBvvI8ifoI37AYek6UDgyhkJOrZ5kOICQpoY5lvEXyPs+r4AFTgReUqygqh4hECd7mpaCoEPKABMLYxIOPTEhRJZURW6C0ch8qMdxrFvpkXHMRail8YJxG6IIHriiA94ya+7E0B+X5Zr+zH-loUGYMBYEQdZ0EAHwALzyjI8oAGSefKCGYEhKHsWhGFYfKrnuV5PkGYhdFviZIXmYeDmAbZoFMf0VkpcBAA8bnJTZkW+TJgXyiEwVmQ0OV5ZZkG+Di-YtPEKD4lmyBCiA2hwIhq49I5PLGKkGpfLsOAOkoqjFCO6AznghAEFAITKAoEAhI8WINJQyjJAQyg4AgvAUPwrSJMkEAAALEMqlD8QAnLIBDHQQJhZk8oaqs+iiusUpxsOcBGgvCoKPM8WrPDU5AzCD5JDSiDBolYdJYt0jLMmy+IXoIADyOBIFAdB4nIyqrDybAsl8WZorqYwrqTcDBGEMBoPxMPoLqbAQDASDIsU9qOs67BdqAcy+J63bepGIABkGuShhG4Z+uMxbzAcsYfF8XQNEmWY4sJIDHe0nRfOkWTqiAP1-ZcoAglcxAAtMaAyLi5i24I9vGqKekCmUozRMUqPmACehdiA-TaKMsTW24sR4vwkjvCKmz6QAwsB2MzAAyjya6YAAslz8o+P1TAMMmg1xsNCc6wrOMNJ0Q55GbHs9LW-FyfKAAKACSbdihaSvAXxo0mC3ahHieCBDHApCCC2GgdF0PRMCdAwKXKxEmnpKtqnPhs9L30Ckrr89fMep79NPYwu27yq6xvvXOIItanxP5+PLP3y7+gz9ng-69N+gQ8R5j1POeGeYxj5mhqn1P+poCzL1XgqRigDmbAJflPR41pv6gPfhAiUqkeIwL0tYUgGxayTl5k6F0yhJyjxyA2X0ODP69mXqNawPsByN1gfpJeSQBisN8HQgoV5lSSBkAPeUbgqht3ynVHeC9pImUIT0CqD4TKMPkXreKFUlHoFUaZTC5klTvmAvJfBrFwFMI4jo-SyprSYTILWBIyhaw8MGPwzAORlRCMYl3eUfERFiOEX4gu7MXEsLYGwzx3iO7d38aIps4iu7iNbqYjKtVw4fw0WVeQ1iVHZPUTRcqBiUwcLvugAMUEQgGSKd+a0+TpFmLlAUnoIRsiFH1IaRmlQohAA)

However, even in the browser it loads more tiles and then the SVG contains more URLs as well as just the two so I probably need to improve the filtering. I wanted to use VegaFusion to see the transformed data but just ran into an error there which I will file in the VF repo.

**Comment by binste** [comment]
I was able to improve the filtering. Here is an example of a chart with only 2 tiles in `tile_list` (I again added a space between the tiles so they are easier to spot). In VS Code, the chart displays almost instantly. [Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAHJQAMlAFYQ2AOxAAaEFBUAzGmTShiNOAHcDmlZhrKGbBhAAScPVjQBmGTI1blVm3YgAdRoAE3p3TwBfSI0QpEwkCDhMCHNEHEwATzQAbWBIgF0YjGcyV1QANk8NQSRMuAAnXNA4hPNlJAQ4NBArQTgAfWEITHUQJIBHBjhlKG7UUESelA0RpAbR1C9xzDYcNAAWaI0ZrRDrfQWQBgbBc104QRCem7uNLJx5kGU2BGskO7FAAe92MTx6INWUAB8xsgkE70ynx6UyQvjo8RopBAxWyVwe4PQ2ShMLQcIRvSRX1R6ISVmxxxAyAaAGtzFBhPtUJgGtMNPAXJtQHAgTgmug+oMIDQAF5wAAE-HlAEYZDjEcj0DRkGRuhoTKFwlcRWKepKBtK5YqVWrGWK2PI4FArCp2TNMI1cl4xDIChoGmw6fMcvxVWovL6Sf1DtVKZqmY1obsmsUeWiINo2A0EM0QEt0EwxtDBFAGLUPT1JtNZnAABReeUHACU6tA2nLHtUqByIELRTUi1S6FeRYBpfLXwA5FhcPgiJLKHsZiMGnBksgcIuGmQCJP5QBqeUyti-AZzGiCA-yye7q+12utBgIShIK-nGgDdsnhpXyUQM92L4TbygApPKf4AQwQFXjee6Hg+8RPpQTBvvI8ifoI37AYek6UDgyhkJOrZ5kOICQpoY5lvEXyPs+r4AFTgReUqygqh4hECd7mpaCoEPKABMLYxIOPTEhRJZURW6C0ch8qMdxrFvpkXHMRail8YJxG6IIHriiA94ya+7E0B+X5Zr+zH-loUGYMBYEQdZ0EAHwALzyjI8oAGSefKCGYEhKHsWhGFYfKrnuV5PkGYhdFviZIXmYeDmAbZoFMf0VkpcBAA8bm1slNnWsqwHeb5MmBfKITBWZDQ5XlBW+EVQkDiA2m6T0MmcdlvkGmEMAWf0alWhpJU+TJyndbWAplJgA0scNAlCf2LTxCg+JZsgQqtXAiGrj0jk8sYqQal8uw4A6SiqMUI7oDOeCEAQUAhMoCgQCEjxYg0lDKMkBDKDgCC8BQ-CtIkyQQAAAsQyqUPxACcsgEKDBAmFmTyhqqz6KK6xSnGw5wEaC8Kgo8zxas8NTkDMZPkidKIMGiVh0li3SMsybL4heggAPI4EgUB0HicjKqsPJsCyXxZmiupjCu4twMEfVoPxdPoLqbAQDASDIsU9qOs67BdqAcy+J63bepGIABkGuShhG4Z+uMxbzAcsYfF8XQNEmWY4sJICg+0nRfOkWTqiAeME5coAglcxAAtMaAyLi5hx4ICfGqKenTa40TFOz5gAnoXYgP02ijLEMduLEeL8JI7wips+kAMLAbzMwAMo8mumAALLa-KPiHUwDDJsdcanQ3vuO3zDSdEOeTh5nPS1vxcnygACgAkmvYoWs7wF8edJgr2oR4nggQxwKQggthoHRdD0TBgwMClysRJp6a7ap30HPS79ApI-b3y+MeU8-Rr5jFTunZUfsP77WcIIWsoCL7gMeLfb4v90DILPAg9+S90BHxPmfU854b5jGAWaSykFfB4NNAWZ+r8FSMUISrYhKCr6PGtNg0h6CKESlUjxWhelrCkA2LWScesnQumUJOU+OQGy+l4Zg3sz9zrWFGLA-B+kn5JAGGo3w8iChXmVJIGQB95RuCqGvBq5cMEP2kiZIRPQaoPhMko+x-t4o1ScegVxplMLmSVO+YC8kBGsXIcojiPj9LKmtJhMg+U0S1h0YMfRmAcjKiMYxLe8o+ImLMcY3JA8NbJNUWwdRGSskb23nk0xTZzFb3MavUJGVqG2L4f7NC0SXFVXkO4mi1UAkphanA9AAYoIhAMoM781pekhPSvNOA-SOrZEKPqQ0MA0BVCTkAA)

For this chart, vl-convert takes 18 seconds to save the chart and the resulting chart does not contain the tiles, only the geoshapes:

![image](https://github.com/vega/vl-convert/assets/23366411/d180d4e0-841a-4b14-ab8b-e11a0885f49b)

**Comment by jonmmease** [comment]
vl-convert uses Vega to generate the SVG and uses resvg to render the SVG to a PNG.  My guess is that any image that shows up in the SVG will be loaded by resvg, even if it doesn't overlap with the view.

This 2-tile example is a good starting point to investigate what vl-convert is doing. I'll let you know what I find!

**Comment by jonmmease** [comment]
Fix in progress in https://github.com/vega/vl-convert/pull/64.  The main issue is that tile.openstreetmap.org requires the requester have a user agent string set.

With this fix, you're final 2-tile example takes well under 1 second to save to PNG (200-400ms on my machine).

The original example takes close to 3 minutes, but it does complete now.

![many_tiles](https://github.com/vega/vl-convert/assets/15064365/cf1051a6-e3eb-4c9b-9997-ab4ff798b5ad)

Hopefully there's a path toward including less tiles in the resulting SVG, because I don't think vl-convert will be able to do this much faster given the architectural limits of how it integrates with resvg (it's not possible to parallelize image downloads).

**Comment by jonmmease** [comment]
Should be fixed in 0.10.3

**Comment by binste** [comment]
🥳 This is awesome, thank you @jonmmease for solving it! I'm relieved that this is fixed.

I was also able to include only the visible tiles in the resulting SVG and now the original spec takes 600ms on my Macbook which is great.
